### PR TITLE
NOISSUE - Use `StatusUnauthorized` instead of `StatusForbidden`

### PR DESCRIPTION
**Pull request by sprql**
Use `401 Unauthorized` instead of `403 Forbidden` for errors related to 
authentication. 

According to [https://tools.ietf.org/html/rfc6750#section-3](https://tools.ietf.org/html/rfc6750#section-3)

### CODE REVIEW

**Comment by sprql** [comment]
As far as I understand, it's better to use 401 instead of 403 for authentication-related errors, or maybe I'm missing something.

**Comment by manuio** [comment]
@sprql sounds good. I would maybe keep forbidden for readers because they use a thing key.

**Comment by sprql** [comment]
@manuio good point, I took this into account and updated the code

**Comment by drasko** [comment]
https://stackoverflow.com/questions/3297048/403-forbidden-vs-401-unauthorized-http-responses

**Comment by manuio** [comment]
@sprql After consulting with maintainers we think that this is correct. Can you update the swaggers, please?

File: authn/api/http/endpoint_test.go
Comment (manuio) [inline authn/api/http/endpoint_test.go]: If you have time I would appreciate that you also fix this logs. It should be `retrieve a token unauthorized` or even `retrieve an unauthorized token`

File: readers/api/transport.go
Comment (drasko) [inline readers/api/transport.go]: This should be kept consistent all over the code. If there is a difference between missing and invalid credentials (AuthN/AuthZ), then it should be detected in other places, and consequently 401/403 (both of them) codes should be used.

**Review by drasko — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (sprql) [reply readers/api/transport.go]: There are 3 cases: 

- Missing token — 401
- Expired token (or maybe invalid by another reason) — 401
- Valid token, but access to some resource with this token is denied — 403

In this example, the current error message ("invalid credentials provided") maybe is a little bit confusing and should be changed to something else (like "permission denied")

Also maybe I should rename `errAccessDenied` to `errPermissionDenied`

File: authn/api/http/endpoint_test.go
Comment (sprql) [reply authn/api/http/endpoint_test.go]: Ok, I'll do

File: readers/api/transport.go
Comment (drasko) [reply readers/api/transport.go]: > There are 3 cases:
>
>Missing token — 401
>Expired token (or maybe invalid by another reason) — 401
>Valid token, but access to some resource with this token is denied — 403

@sprql exactly - well understood. Name of the error does not bother me much - `errAccessDenied` is OK and it relates to `403` rather than `401`, i.e. it is self-documenting enough.

What I wanted to check with you is did you cover all the code code base and identified where one of these 3 cases might appear and well differentiated between them - because here I can see `403` practically appearing just at one place in the whole Mainflux, according to your changes.

I did not review whole PR in details - it's bit difficult to review, this is why I am asking (looks to me you know your way areound Mainflux now and you know what you are doing, which is very good).

Anyway - thanks for this PR, it is very nice and think that you are really on a good track here for a nice improvement. Just please do not forget to update Swagger files as well.

@nmarcetic and @dusanb94 should review this one as well.

Comment (dborovcanin) [reply readers/api/transport.go]: I agree with this approach. This will later translate to handling errors from AuthN and AuthZ services (at the moment, only AuthN exists), where errors from AuthN will cause `Forbidden`, and errors from AuthZ will be `Unauthorized`.

Comment (drasko) [reply readers/api/transport.go]: Great, we have general alignment of maintainers then.

@sprql please double-check the PR, ad final touches prepare it for meging. Please do not forget to add Swagger changes.

After that PR will be merged.

Comment (sprql) [reply readers/api/transport.go]: @drasko, currently, the code mainly uses the `errUnauthorizedAccess`, both for AuthN & AuthZ errors, therefore, we will need to enter a new error type in these places. 

One option is to use error `errUnauthorizedAccess` for AuthN, and enter `errPermissionDenied` for AuthZ. But 'errUnauthorizedAccess' is misleading because of its name. 

Alternatively, we can enter `errUnauthenticated' for AuthN errors (like in gRPC).

In other words, the following options are possible:
`errUnauthorizedAccess` (401), `errPermissionDenied` (403)
or `errUnauthenticated` (401), `errUnauthorizedAccess` (403)
or `errUnauthenticated` (401), `errPermissionDenied` (403)

What do you think about it?

Comment (drasko) [reply readers/api/transport.go]: Good question... I would avoid option 2, as this can be misleading (`errUnauthorizedAccess` might be related to 401 when here should be related to 403).

Between options 1 and 3, probably 3 is the least ambigous, although we could probably understant 1 as well now that we know what this is about...

I would personally vote for 3, to remove any abiguity, but I am also interested what others think. @manuio @nmarcetic @dusanb94 @mteodor @blokovi

Comment (manuio) [reply readers/api/transport.go]: I also vote for 3

Comment (sprql) [reply readers/api/transport.go]: I updated PR using option 3

Comment (drasko) [reply readers/api/transport.go]: @sprql please do ot hate me, but after looking at PR I realized that actually option 2 would be more apropriate for Mainflux.

Actually, in Mainflux we have AuthN and AuthZ services, and I think that we should align errors to these functionalities, and not to outdated and missleading HTTP status names.

So I propose option 2, with a slight change: instad of `errUnauthorizedAccess` for `403` just use name `errUnauthorized`, to align it to it's 401 counterpart (`errUnauthenticated`).

Sorry for this, but I think now that everything is almost done, it will mostly be find-replace job :).

Comment (sprql) [reply readers/api/transport.go]: @drasko I've updated, could you look at this again, I also updated error messages

File: authn/api/http/transport.go
Comment (dborovcanin) [inline authn/api/http/transport.go:107]: Should there be 2 cases (in this particular order):
```go
case errors.Contains(err, authn.ErrUnauthenticated):
    w.WriteHeader(http.StatusUnauthorized)
case errors.Contains(err, authn.ErrUnauthorized):
    w.WriteHeader(http.StatusForbidden)
```
The service itself won't return ErrUnauthenticated, but our JWT implementation (or transport layer in the case `key==""`) might.

File: authn/swagger.yaml
Comment (dborovcanin) [inline authn/swagger.yaml:57]: If we include both 401 and 403 (see my previous comment), ew should update API docs accordingly.

File: bootstrap/api/transport.go
Comment (dborovcanin) [inline bootstrap/api/transport.go:266]: The same remark from Authn service for adding `Unauthorized` case applies here.
Not to repeat myself and spam this thread, please take a look at all the other services and consider adding both cases both in the transport layer (`svc/api/transport.go`) and API docs (Swagger files).

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Comment by dborovcanin** [comment]
@sprql Please don't address my comments yet. @mainflux/maintainers Need to agree on this. We'll do that internally, bring the final decision and let you know. We apologize for our back and forth request changes. I find both 401 and 403 code verbose values confusing and we had an internal misunderstanding between team members.

**Comment by sprql** [comment]
@dusanb94, do you have any updates?

**Comment by dborovcanin** [comment]
> @dusanb94, do you have any updates?

Unfortunately, not yet. Mainflux team is overloaded with some other urgent tasks, but we will get back to this PR as soon as we finish them.

**Comment by drasko** [comment]
I would like that we can back to this next week, right after `0.11` release.

**Comment by drasko** [comment]
@sprql can you please resolve the conflicts and update the branch. This starts to look good, and I would like that we proceed with this big PR.

@manuio @nmarcetic @dusanb94 please review after update.

**Review by blokovi — DISMISSED** [review]

File: bootstrap/api/transport.go
Comment (sprql) [reply bootstrap/api/transport.go:266]: > @sprql Please don't address my comments yet. @mainflux/maintainers Need to agree on this. We'll do that internally, bring the final decision and let you know. 

@dusanb94  Have you made a decision on this? Should I add this cases?

Comment (dusanb94) [reply bootstrap/api/transport.go:266]: We should probably add these cases, but please don't address this comment yet. We didn't decide what to do with this PR yet, and there is no point that you continue working on this before we give you feedback.

Comment (dborovcanin) [reply bootstrap/api/transport.go:266]: @sprql Let's do it like that. Please update the branch, add these cases, and let's review and merge.

**Comment by mteodor** [comment]
`Vault` API has only `403` code 

```
204 - Success, no data returned.
400 - Invalid request, missing or invalid data.
403 - Forbidden, your authentication details are either incorrect, you don't have access to this feature, or - if CORS is enabled - you made a cross-origin request from an origin that is not allowed to make such requests.
404 - Invalid path. This can both mean that the path truly doesn't exist or that you don't have permission to view a specific path. We use 404 in some cases to avoid state leakage.

```
https://www.vaultproject.io/api-docs#http-status-codes

**Comment by nmarcetic** [comment]
@sprql I am checking PR , for example 

```
case errors.Contains(err, authn.ErrUnauthenticated):
    w.WriteHeader(http.StatusUnauthorized)
case errors.Contains(err, authn.ErrUnauthorized):
    w.WriteHeader(http.StatusForbidden)
```

 Why not follow go HTTP lib and use Forbidden for 403 (as its status code Forbidden by spec) and Unautrhozed for 401 (as its status code for Unauthorized by spec)?
You added ErrUnauthenticated for 401, why "custom "naming?  And from the same link, the [spec you are addressing](https://tools.ietf.org/html/rfc6750#section-3) it says 

```
  For example, in response to a protected resource request without
   authentication:

     HTTP/1.1 401 Unauthorized
     WWW-Authenticate: Bearer realm="example"


```

IMHO `ErrUnauthorized` makes more sense than `ErrUnauthenticated` for status code 401. What do you think?

**Comment by sprql** [comment]
> Why not follow go HTTP lib and use Forbidden for 403 (as its status code Forbidden by spec) and Unautrhozed for 401 (as its status code for Unauthorized by spec)?

@nmarcetic 
I suggested 3 different options https://github.com/mainflux/mainflux/pull/1140#discussion_r418028870

**Comment by nmarcetic** [comment]
> > Why not follow go HTTP lib and use Forbidden for 403 (as its status code Forbidden by spec) and Unautrhozed for 401 (as its status code for Unauthorized by spec)?
> 
> @nmarcetic
> I suggested 3 different options [#1140 (comment)](https://github.com/mainflux/mainflux/pull/1140#discussion_r418028870)

@sprql I saw that you voted, but too late, sorry for that. @manuio explained to me but I already left my comment. 
I still think it's confusing and that option 3 is worst, but we have consensus already :) 
So let's see if we can approve and merge this @drasko @dusanb94 You must approve (there are your remarks blocking merge)
Also @sprql can you try to update, a branch is out of date. Thanks.

**Comment by nmarcetic** [comment]
Can we update this branch and resolve conflicts @sprql

**Comment by sprql** [comment]
@nmarcetic I've updated branch

**Comment by darkodraskovic** [comment]
Finally, we've decided to use `http.StatusUnauthorized` for the authentication and `http.StatusForbidden` for authorization.

**Pull request closed without merging**
